README
Data::Phrasebook
================

DESCRIPTION

  Abstract your queries!

This distribution enables the framework for abstracting common or locale data
out of your code into a datastore. The default is within a plain text file,
but more complex datastores can be intrigrated via the use of other 
Data::Pharsebook::Loader distributions (eg YAML, XML, INI, DBI).

DEPENDENCIES

The distribution requires the following modules:

  Module::Pluggable   >= 2.5

For debugging purposes, the following modules are required:

  Data::Dumper

For testing purposes, the following modules are required:

  Test::More          >= 0.47

For testing purposes, the following modules are desireable, but not essential:

  Test::Pod           >= 1.00
  Test::Pod::Coverage >= 0.08
  Pod::Coverage
  Test::CPAN::Meta    >= 0.12

INSTALLATION

To install this module type the following:

 perl Makefile.PL
 make
 make test
 make install

For more detailed installation instructions, see INSTALL.

USAGE

Read the pod documentation in the distrubtion files.

CHANGES

For a complete list of changes, see the Changes file.

SUPPORT

If you spot a bug or are experiencing difficulties that are not explained 
within the POD documentation, please send an email to barbie@cpan.org or 
submit a bug to the RT system (http://rt.cpan.org/). It would help greatly 
if you are able to pinpoint problems or even supply a patch. 

Fixes are dependent upon their severity and my availability. Should a fix not
be forthcoming, please feel free to (politely) remind me.

AUTHOR

Original author: Iain Campbell Truskett (16.07.1979 - 29.12.2003)

Maintainer: Barbie <barbie@cpan.org> for Miss Barbell Productions 
http://www.missbarbell.co.uk.

COPYRIGHT AND LICENSE

  Copyright (C) 2004-2013 Barbie for Miss Barbell Productions.
  Copyright (C) 2003 Iain Truskett.

  This distribution is free software; you can redistribute it and/or
  modify it under the Artistic Licence v2.


Changes
Changes log for Data-Phrasebook
===============================

        - fixed license fields in META.json to be lists.
        - extended test suite.
        - added repository link to META.yml.

0.34    2013-05-27
        - change file dates changed to meet W3CDTF standards.
        - clarified licensing text.

0.33    2012-07-29
        - fixed META.json test file.

0.32    2012-07-29
        - removed License files from distro.
        - Artistic License v2 now sole license.
        - removed DSLIP info.
        - spelling fixes. (thanks to Nicholas Bamber, Florian Schlichting & 
          Gregor Herrmann).
        - added minimum perl version (5.006).
        - reworked Makefile.PL for clarity.
        - updated META.yml testing.
        - added Carp as a prerequisite.
        - implemented Perl::Critic suggestions.

0.31    2010-08-31
        - added blank_args as a parameter to delimiters().

0.30    2010-08-20
        - removed 5.006 dependency
        - upgrade to Test::CPAN::Meta.
        - added META.json
        - minor code clean
        - now released under the Artistic License v2.

0.29    2007-03-02
        - RT#25089 - if delimiters are set via new(), the default settings no
          longer reset in Generic.pm. (Thanks to Javier Gonzalez).

0.28    2007-03-01
        - Fixed META.yml typos.
        - Added META.yml test script.
        - Removed distribution test, as it can't handle optional 'use'

0.27    2007-02-22
        - Handwritten META.yml to conform to current specs.
        - Changed all instances of LICENCE to LICENSE as *apparently*, to all
          non-British users of this module, this a bug! I beg to differ -
          http://dictionary.reference.com/search?q=licence
        - Added an examples directory. Nothing in it yet, except a request
          for interesting scripts.
        - Added missing checks for debug calls
        - Ensure DP::Generic::loader() returns a value.
        - Ensure DP::Loader classes know their own class.
        - More tests!
        - Devel::Cover - 90.4%

0.26    2006-09-28
        - Patch supplied from Oliver Gorwits to support the ability of passing
          an array reference for the list of dictionaries. 
        - Fixed a bug in new, where the attributes were not getting set
          correctly. (Oliver Gorwits)
        - Fixed POD to remove reference to obsolete method.
        - more tests
        - Devel::Cover - 89.9%

0.25    2005-11-21
        - Added support for multiple dictionaries.
        - Finally added the Module::Pluggable::search_path functionality
          (after Simon implemented my patch 9 months ago! .. I was busy m'kay)
        - The query() method in SQL phrasebook class, can now support 
          replacement of parameters as well as binding of parameters.
          (Thanks to Tielman De Villiers for the idea and the code examples)

0.24    2005-09-22
        - Added sorting to keywords() for Text.pm loader.

0.23    2005-08-09
        - Fixed fetch/execute code in Query.pm (Tielman De Villiers)
        - provide support for returning the list of keywords available.

0.22    2005-06-13
        - POD suggestions (Jon Allen).
        - delimiters now a generic class behaviour.
        - provide support for returning the list of dictionaries available
          (suggested by Jon Allen).

0.21    2005-05-13
        - More POD to explain Classes and Dictionaries.
        - file & dictionary reloading.
        - DESTROY bug fix in SQL/Query.pm (Ray Miller).
        - Aligned versions, just to make life easy.

0.20    2005-03-03
        - POD updates.

0.19    2005-03-02
        - Removed Test::MockObject from tests. No longer needed.

0.18    2005-03-01
        - Name change from Phrasebook to Data::Phrasebook.
        - Loader now uses Module::Pluggable due to ease of use and the fact
          that Class::MethodMaker now regularly fails on Win32.
        - Removed Class::WhiteHole.
        - Major documentation overhaul.

0.17    2004-02-09
        - New maintainer - Barbie - as of 31 Jan 2003.
        - Default loader, YAML, now a separate distribution.
        - Default loader is now plain text. This was done to minimise the
          required modules, and to enable the appropriate datastore for
          loading to be install separately.
        - dictionary/section setting now implemented as:
          Phrasebook->new( dict => 'NEW' );

0.16    2003-08-03
        - Fixed execution logic.
        - Take the right number of arguments off the stack.
        - A bit more elegant at getting the method's name.
        - Query refactored a bit.
        - Documentation fixed regarded some implementations.
        - No tabs.

0.15    2003-07-27
        - Abstracted the loader aspect.
        - Changed default Plain delimiters to match TT.
        - Added facility to let one change delimiters.
        - Added test for above.
        - One can pass an explicit class name to 'new'.
        - Added the documentation.
        - Set loader defaults.
        - Ensure all .pm files compile well.
        - Added strict and fatal warnings to everything.
        - Removed unused imports.
        - Cache delegations.
        - Add strict+warnings.
        - Fix variable name.

0.14    2003-07-25
        - Added versions to everything.
        - Added WhiteHole to generic.pm to catch wayward methods.

0.13    2003-07-25
        - Initial release.


lib/Data/Phrasebook.pm
package Data::Phrasebook;
use strict;
use warnings FATAL => 'all';
use base qw( Data::Phrasebook::Debug );
use Carp qw( croak );

use vars qw($VERSION);
$VERSION = '0.34';

=head1 NAME

Data::Phrasebook - Abstract your queries!

=head1 SYNOPSIS

    use Data::Phrasebook;

    my $q = Data::Phrasebook->new(
        class  => 'Plain',
        loader => 'Text',
        file   => 'phrases.txt',
    );

    # simple keyword to phrase mapping
    my $phrase = $q->fetch($keyword);

    # keyword to phrase mapping with parameters
    $q->delimiters( qr{ \[% \s* (\w+) \s* %\] }x );
    my $phrase = $q->fetch($keyword,{this => 'that'});

=head1 DESCRIPTION

Data::Phrasebook is a collection of modules for accessing phrasebooks from
various data sources.

=head1 PHRASEBOOKS

To explain what phrasebooks are it is worth reading Rani Pinchuk's
(author of L<Class::Phrasebook>) article on Perl.com:

L<http://www.perl.com/pub/a/2002/10/22/phrasebook.html>

Common uses of phrasebooks are in handling error codes, accessing databases
via SQL queries and written language phrases. Examples are the mime.types
file and the hosts file, both of which use a simple phrasebook design.

Unfortunately Class::Phrasebook is a complete work and not a true class
based framework. If you can't install XML libraries, you cannot use it.
This distribution is a collaboration between Iain Truskett and myself to
create an extendable and class based framework for implementing phrasebooks.

=head1 CLASSES

In creating a phrasebook object, a class type is required. This class defines
the nature of the phrasebook or the behaviours associated with it. Currently
there are two classes, Plain and SQL.

The Plain class is the default class, and allows retrieval of phrases via the
fetch() method. The fetch() simply returns the phrase that maps to the given
keyword.

The SQL class allows specific database handling. Phrases are retrieved via the
query() method. The query() method internally retrieves the SQL phrase, then
returns the statement handler object, which the user can then perform a
prepare/execute/fetch/finish sequence on. For more details see
Data::Phrasebook::SQL.

=head1 CONSTRUCTOR

=head2 new

The arguments to new depend upon the exact class you're creating.

The default class is C<Plain> and only requires the Loader arguments. The
C<SQL> class requires a database handle as well as the Loader arguments.

The C<class> argument defines the object class of the phrasebook and the
behaviours that can be associated with it. Using C<Foobar> as a fake class,
the class module is searched for in the following order:

=over 4

=item 1

If you've subclassed C<Data::Phrasebook>, for example as C<Dictionary>,
then C<Dictionary::Foobar> is tried.

=item 2

If that failed, C<Data::Phrasebook::Foobar> is tried.

=item 3

If B<that> failed, C<Foobar> is tried.

=item 4

If all the above failed, we croak.

=back

This should allow you some flexibility in what sort of classes
you use while not having you type too much.

For other parameters, see the specific class you wish to instantiate.
The class argument is removed from the arguments list and the C<new>
method of the specified class is called with the remaining arguments.

=cut

sub new {
    my $class = shift;
    my %args = @_;

    my $debug = delete $args{debug} || 0;
    $class->debug($debug);

    if($debug) {
		$class->store(3,"$class->new IN");
		$class->store(4,"$class->new args=[".$class->dumper(\%args).']');
	}

    my $sub = delete $args{class} || 'Plain';
    if (eval "require ${class}::$sub") {
        $sub = $class."::$sub";
    } elsif (eval "require Data::Phrasebook::$sub") {
        $sub = "Data::Phrasebook::$sub";
    } elsif (eval "require $sub") {
        # it's a module by itself
    } else {
        croak "Could not find appropriate class for '$sub': [$@]";
    }

    $class->store(4,"$class->new sub=[$sub]")	if($class->debug);

    return $sub->new( %args );
}

1;

__END__

=head1 DELIMITERS

Delimiters allow for variable substitution in the phrase. The default style
is ':variable', which would be passed as:

    $q->delimiters( qr{ :(\w+) }x );

As an alternative, a Template Toolkit style would be passed as:

    $q->delimiters( qr{ \[% \s* (\w+) \s* %\] }x );

=head1 DICTIONARIES

=head2 Simple Dictionaries

Data::Phrasebook supports the use of dictionaries. See the specific Loader
module to see how to implement the dictionary within your phrasebook. Using
Data::Phrasebook::Loader::Ini as an example, the dictionary might be laid out
as:

  [Stuff]
  language=Perl
  platform=Linux

  [Nonsense]
  platform=Windows

The phrasebook object is then created and used as:

  my $q = Data::Phrasebook->new(
    class  => 'Plain',
    loader => 'Ini',
    file   => 'phrases.ini',
    dict   => 'Nonsense',
  );

  my $language = $q->fetch('language');	# retrieves 'Perl'
  my $platform = $q->fetch('platform');	# retrieves 'Windows'

The former is from the default (first) dictionary, and the second is from the
named dictionary ('Nonsense'). If a phrase is not found in the named dictionary
an attempt is made to find it in the default dictionary. Otherwise undef will
be returned.

Once a dictionary or file is specified, changing either requires reloading. As
this is done at the loader stage, we need to let it know what it needs to
reload. This can be done with the either (or both) of the following:

  $q->file('phrases2.ini');
  $q->dict('Stuff');

A subsequent fetch() will then reload the file and dictionary, before
retrieving the phrase required. However, a reload only takes place if both the
file and the dictionary passed are not the ones currently loaded.

=head2 Multiple Dictionaries

As of version 0.25, the ability to provide prescendence over multiple
dictionaries for the same phrasebook. Using Data::Phrasebook::Loader::Ini
again as an example, the phrasebook might be laid out as:

  [AndTheOther]
  language=Perl
  platform=Linux
  network=LAN

  [That]
  platform=Solaris
  network=WLAN

  [This]
  platform=Windows

The phrasebook object is then created and used as:

  my $q = Data::Phrasebook->new(
    class  => 'Plain',
    loader => 'Ini',
    file   => 'phrases.ini',
    dict   => ['This','That','AndTheOther'],
  );

  my $language = $q->fetch('language');	# retrieves 'Perl'
  my $platform = $q->fetch('platform');	# retrieves 'Windows'
  my $network  = $q->fetch('nework');	# retrieves 'WLAN'

The first dictionary, if not specified and supported by the Loader module, is
still used as the default dictionary.

The dictionaries can be specified, or reordered, using the object method:

  $q->dict('That','AndTheOther','This');

A subsequent reload will occur with the next fetch call.

=head1 DEDICATION

Much of the code for the original class framework is from Iain's original code.
My code was much simpler and was tied to using just an INI data source. Merging
all the ideas and code together we came up with this distribution.

Unfortunately Iain died in December 2003, so he never got to see or play
with the final working version. I can only thank him for his thoughts and
ideas in getting this distribution into a state worthy of release.

  Iain Campbell Truskett (16.07.1979 - 29.12.2003)

=head1 SEE ALSO

L<Data::Phrasebook::Plain>,
L<Data::Phrasebook::SQL>,
L<Data::Phrasebook::SQL::Query>,
L<Data::Phrasebook::Debug>,
L<Data::Phrasebook::Generic>,
L<Data::Phrasebook::Loader>,
L<Data::Phrasebook::Loader::Text>,
L<Data::Phrasebook::Loader::Base>.

=head1 SUPPORT

Please see the README file.

=head1 AUTHOR

  Original author: Iain Campbell Truskett (16.07.1979 - 29.12.2003)
  Maintainer: Barbie <barbie@cpan.org> since January 2004.
  for Miss Barbell Productions <http://www.missbarbell.co.uk>.

=head1 COPYRIGHT AND LICENSE

  Copyright (C) 2003 Iain Truskett.
  Copyright (C) 2004-2013 Barbie for Miss Barbell Productions.

  This distribution is free software; you can redistribute it and/or
  modify it under the Artistic License v2.

=cut
